test/env/env.go

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2021 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package env
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	webhooktesting "github.com/jetstack/cert-manager/cmd/webhook/app/testing"
+	"github.com/onsi/ginkgo"
+	ginkgoconfig "github.com/onsi/ginkgo/config"
+	"github.com/onsi/ginkgo/reporters"
+	"github.com/onsi/gomega"
+	extapi "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+
+	cmpapi "github.com/cert-manager/policy-approver/apis/policy/v1alpha1"
+)
+
+func init() {
+	// Turn on verbose by default to get spec names
+	ginkgoconfig.DefaultReporterConfig.Verbose = true
+	// Turn on EmitSpecProgress to get spec progress (especially on interrupt)
+	ginkgoconfig.GinkgoConfig.EmitSpecProgress = true
+	// Randomize specs as well as suites
+	ginkgoconfig.GinkgoConfig.RandomizeAllSpecs = true
+
+	wait.ForeverTestTimeout = time.Second * 60
+
+	// Since we are going to be creating CRDs in the API server, we need to
+	// register these types.
+	utilruntime.Must(extapi.AddToScheme(cmpapi.GlobalScheme))
+}
+
+// RunSuite runs a Ginkgo test suite, and writes the results to the artefacts
+// directory. The artifacts directory may be overridden with `ARTIFACTS`
+// environment variable.
+func RunSuite(t *testing.T, suiteName, artifactsDir string) {
+	gomega.RegisterFailHandler(ginkgo.Fail)
+
+	if path := os.Getenv("ARTIFACTS"); len(path) > 0 {
+		artifactsDir = path
+	}
+
+	junitReporter := reporters.NewJUnitReporter(filepath.Join(
+		artifactsDir,
+		fmt.Sprintf("junit-go-%s.xml", suiteName),
+	))
+
+	ginkgo.RunSpecsWithDefaultAndCustomReporters(t, suiteName, []ginkgo.Reporter{junitReporter})
+}
+
+// RunControlPlane runs a local API server and makes it ready for running tests
+// against. Also runs the cert-manager webhook for operating over cert-manager
+// resources. Expects CRD directories to both cert-manager, as well as
+// policy-approver. This *MUST* be provided.
+// Returns a controller-runtime envtest which is ready to be run against.
+func RunControlPlane(t *testing.T, crdDirs ...string) *envtest.Environment {
```

Comment:
> So in the e2e tests we don't run an actual cert-manager, we only create the CRDs and spin up the webhook because the API server will call call webhook endpoints when custom resources are applied 👍🏼

Yep :slightly_smiling_face: 

Yes, ideally this API isn't going to change, or at least in a way in which policy would no longer work. We should be in a state where any cert-manager policy should work with any cert-manager version v1.3+